teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,639 @@
+pr-produce-platforms-and-browsers-8 - "path/to/report/windows-firefox"
+```
+
+### Goals
+
+The main goal of this TEP is to enable executing a `matrixed` `pipelineTask` that can emit `result` and support
+consuming all of the `results` that were produced in another `pipelineTask` as long as the type of `result` emitted
+was `string`.
+
+### Non-Goals
+
+The following are out of scope for this TEP:
+
+1. Support a matrixed `pipelineTask` that produces `result` of type `array` or `object` that are then consumed by
+   another `pipelineTask`.
+  - A matrixed `pipelineTask` can support `result` of any type as long as they aren’t consumed by
+    another `pipelineTask`. However, only matrixed `pipelineTask` that produces `result` of type `string` can be
+    consumed by another `pipelineTask`.
+2. Support consuming a specific instance or combination(s) of `result` produced by a fanned out `pipelineTask`.
+  - At this time, we propose only supporting whole `array` `result` replacements from a matrixed `pipelineTask`
+
+### Background
+
+Consuming `results` from previous `taskRuns` or `runs` in a `matrix`, which would dynamically generate `taskRuns` from
+the fanned out `results`, is currently supported for string `results`, array `results`, and object `results`. Note that
+the underlying `results` type in each must be string.
+
+#### String Results
+
+String `results` is a stable feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>)` and can
+be passed into the `matrix` from previous `taskRuns`.
+
+```yaml
+tasks:
+...
+- name: task-1
+  taskRef:
+    name: task-1
+  matrix:
+    - name: values
+      value:
+        - $(tasks.task-1.results.foo) # string
+```
+
+#### Array Results
+
+Array `results` is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`. String
+replacements from arrays are supported through array indexing and are referred to
+as `$(tasks.<pipelinetask-name>.results.<result-name>[i])`  where i is the index. Array `results` from
+previous `taskRuns` can be passed into the Matrix:
+
+```yaml
+tasks:
+...
+- name: task-2
+  taskRef:
+    name: task-2
+  matrix:
+    - name: values
+      value: $(tasks.task-4.results.bar[*]) # array
+```
+
+#### Object Results
+
+Object `results` is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>.[*])`.
+String replacements from objects are supported and are referred to as `$(tasks.<pipelinetask-name>.results.<result-name>
+.key)`  where key is the object key. Strings from Object `results` from previous `taskRuns` can be passed into
+the `matrix` tasks:
+
+```yaml
+...
+- name: task-3
+  taskRef:
+    name: task-3
+  matrix:
+    - name: values
+      value: $(tasks.task-4.results.rad.key) # string replacement from object result
+```
+
+### Requirements
+
+1. A `matrix` `pipelineTask` can only produce `result` of type `string` if the `results` are being consumed by
+   another `pipelineTask`.
```

Comment:
clarifying the requirement -- i believe this was the intent; the current wording may indicate that array and object results are not allowed even if they are not consumed 

```suggestion
1. A `matrix` `pipelineTask` can produce `results` any type, but only `results` of type `string` can being consumed by another `pipelineTask`.
```